internal/juju/jaas.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package juju
+
+import (
+	"github.com/canonical/jimm-go-sdk/v3/api"
+	"github.com/canonical/jimm-go-sdk/v3/api/params"
+	jujuapi "github.com/juju/juju/api"
+
+	"github.com/juju/errors"
+)
+
+type jaasClient struct {
+	SharedClient
+	getJaasApiClient func(jujuapi.Connection) JaasAPIClient
+}
+
+func newJaasClient(sc SharedClient) *jaasClient {
+	return &jaasClient{
+		SharedClient: sc,
+		getJaasApiClient: func(conn jujuapi.Connection) JaasAPIClient {
+			return api.NewClient(conn)
+		},
+	}
+}
+
+func (jc *jaasClient) AddRelations(tuples []params.RelationshipTuple) error {
+	conn, err := jc.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+	cl := jc.getJaasApiClient(conn)
+	req := params.AddRelationRequest{
+		Tuples: tuples,
+	}
+	return cl.AddRelation(&req)
```

Comment:
what error is returned in case the relation already exists? should this be treated as an error? i think it should not and nil should be returned in such a  case.

Reply:
That's defined by the server side and how we implement idempotency. Currently the API is not idempotent, it's also tricky to make it idempotent because it's a bulk call so if one tuple in the array already exists, I believe the entire transaction is aborted.
But I have thought on this and I think it's okay as it currently is.

Reply:
Thought: If you allow for a bulk call, the API should return errors for each element of the bulk call and an error for the call itself.  This would allow for ignoring AlreadyExists errors and the NotFound below. 

There is a set of github issues related to failures of AlreadyExists and NotFound (on delete) which need to be fixed.

Reply:
Please link to a Jira card to address this if the work will be done later. Allows for follow up here.

Reply:
I'll resolve this without a ticket. It's not currently planned.